File: AI/src/Graph.java
import java.util.*;

/**
 * Created by oc on 2017/10/12.
 */
public class Graph {

    public static final int ROAD = 0;
    public static final int WALL = 1;


    public static final int RED_JEWEL = 11;
    public static final int BLUE_JEWEL = 12;
    public static final int GREEN_JEWEL = 13;

    public static final int YELLOW_KEY = 15;
    public static final int BLUE_KEY = 16;
    public static final int RED_KEY = 17;

    public static final int RED_POTION = 18;
    public static final int BLUE_POTION = 19;
    public static final int YELLOW_POTION = 20;
    public static final int GREEN_POTION = 21;

    public static final int DOOR_YELLOW = 81;
    public static final int DOOR_BLUE = 82;
    public static final int DOOR_RED = 83;

    public static final int UPSTAIR = 87;
    public static final int DOWNSTAIR = 88;

    public static final int MONSTER_BOUND = 101;
    public static final int BOSS_INDEX = 199;


    int floor, row, col;
    int[][][] map;
    // up 0,1  down 2,3
    int[][] stair;

    int bossId=-1;

    int p_atk, p_def, p_mdef, p_red, p_blue, p_yellow, p_green;
    HashMap<Integer, Monster> monsterMap;

    ArrayList<Node> list;

    public Graph(Scanner scanner) {
        list=new ArrayList<>();
        floor=scanner.nextInt(); row=scanner.nextInt(); col=scanner.nextInt();
        map=new int[floor][row][col];
        stair=new int[floor][4];
        for (int i=0;i<floor;i++) stair[i]=new int[]{-1,-1,-1,-1};

        for (int i=0;i<floor;i++) {
            for (int j=0;j<row;j++) {
                for (int k=0;k<col;k++) {
                    map[i][j][k]=scanner.nextInt();
                }
            }
        }

        // points
        p_atk=scanner.nextInt();
        p_def=scanner.nextInt();
        p_mdef=scanner.nextInt();
        p_red=scanner.nextInt();
        p_blue=scanner.nextInt();
        p_yellow=scanner.nextInt();
        p_green=scanner.nextInt();

        // Monsters
        monsterMap=new HashMap<>();
        int num=scanner.nextInt();
        for (int i=0;i<num;i++) {
            int id=scanner.nextInt();
            monsterMap.put(id, new Monster(id, scanner.nextInt(), scanner.nextInt(), scanner.nextInt(), scanner.nextInt()));
        }

        // Initial Node
        int hp=scanner.nextInt(), atk=scanner.nextInt(), def=scanner.nextInt(), mdef=scanner.nextInt(),
                yellow=scanner.nextInt(), blue=scanner.nextInt(), red=scanner.nextInt(),
                floor=scanner.nextInt(), x=scanner.nextInt(), y=scanner.nextInt();

        Node node=new Node(0, new Hero(hp, atk, def, mdef, yellow, blue, red), null, null, 0,
                floor, x, y);
        list.add(node);

        buildMap();

        mergeNode();

        // set id
        for (int i=0;i<list.size();i++) {
            list.get(i).setId(i);
            if (list.get(i).type==BOSS_INDEX)
                bossId=i;
        }

    }

    public void buildMap() {
        for (int i=0;i<floor;i++) {
            for (int j=0;j<row;j++) {
                for (int k=0;k<col;k++) {
                    Node node=null;
                    if (map[i][j][k]==UPSTAIR) {
                        stair[i][0]=j; stair[i][1]=k;
                    }
                    if (map[i][j][k]==DOWNSTAIR) {
                        stair[i][2]=j; stair[i][3]=k;
                    }
                    if (map[i][j][k]==YELLOW_KEY)
                        node=new Node(map[i][j][k], null, null,
                                new Item(0,0,0,0,1,0,0),
                                0,i,j,k);
                    if (map[i][j][k]==BLUE_KEY)
                        node=new Node(map[i][j][k], null, null,
                                new Item(0,0,0,0,0,1,0),
                                0,i,j,k);
                    if (map[i][j][k]==RED_KEY)
                        node=new Node(map[i][j][k], null, null,
                                new Item(0,0,0,0,0,0,1),
                                0,i,j,k);
                    if (map[i][j][k]==RED_JEWEL)
                        node=new Node(map[i][j][k], null, null,
                                new Item(0,p_atk,0,0,0,0,0),
                                0,i,j,k);
                    if (map[i][j][k]==BLUE_JEWEL)
                        node=new Node(map[i][j][k], null, null,
                                new Item(0,0,p_def,0,0,0,0),
                                0,i,j,k);
                    if (map[i][j][k]==GREEN_JEWEL)
                        node=new Node(map[i][j][k], null, null,
                                new Item(0,0,0,p_mdef,0,0,0),
                                0,i,j,k);
                    if (map[i][j][k]==RED_POTION)
                        node=new Node(map[i][j][k], null, null,
                                new Item(p_red,0,0,0,0,0,0),
                                0,i,j,k);
                    if (map[i][j][k]==BLUE_POTION)
                        node=new Node(map[i][j][k], null, null,
                                new Item(p_blue,0,0,0,0,0,0),
                                0,i,j,k);
                    if (map[i][j][k]==YELLOW_POTION)
                        node=new Node(map[i][j][k], null, null,
                                new Item(p_yellow,0,0,0,0,0,0),
                                0,i,j,k);
                    if (map[i][j][k]==GREEN_POTION)
                        node=new Node(map[i][j][k], null, null,
                                new Item(p_green,0,0,0,0,0,0),
                                0,i,j,k);

                    if (map[i][j][k]==DOOR_YELLOW)
                        node=new Node(map[i][j][k], null, null, null, 1, i,j,k);
                    if (map[i][j][k]==DOOR_BLUE)
                        node=new Node(map[i][j][k], null, null, null, 2, i,j,k);
                    if (map[i][j][k]==DOOR_RED)
                        node=new Node(map[i][j][k], null, null, null, 3, i,j,k);
                    if (map[i][j][k]>=MONSTER_BOUND) {
                        Monster monster=monsterMap.get(map[i][j][k]);
                        if (monster==null) continue;
                        node=new Node(map[i][j][k], null, monster, null, 0, i, j, k);
                    }

                    if (node!=null)
                        list.add(node);
                }
            }
        }

        // build graph
        int len=list.size();
        for (int i=0;i<len;i++) {
            for (int j=i+1;j<len;j++) {
                Node n1=list.get(i), n2=list.get(j);
                if (isLinked(n1.f, n1.x, n1.y, n2.f, n2.x, n2.y)) {
                    n1.addNode(n2);
                    n2.addNode(n1);
                }
            }
        }
    }


    private void mergeNode() {
        for (int i=1;i<list.size();i++) {
            Node n1=list.get(i);
            for (int j=i+1;j<list.size();j++) {
                Node n2=list.get(j);
                if (shouldMerge(n1, n2)) {
                    n1.merge(n2);
                    list.remove(j);
                    mergeNode();
                    return;
                }
            }
        }
    }

    private boolean shouldMerge(Node n1, Node n2) {
        if (n1.item!=null || n2.item!=null) return false;
        if (n1.type==BOSS_INDEX || n2.type==BOSS_INDEX) return false;
        if (!n1.linked.contains(n2) || !n2.linked.contains(n1)) return false;
        for (Node node: n2.linked)
            if (n1.linked.contains(node))
                return false;
        for (Node u: new Node[] {n1, n2}) {
            Node v=u==n1?n2:n1;
            for (Node x: u.linked) {
                for (Node y: u.linked) {
                    if (x.equals(y) || x.equals(v) || y.equals(v)) continue;
                    if (!x.linked.contains(y) || !y.linked.contains(x)) return false;
                }
            }
        }
        return true;
    }

    private boolean isLinked(int f1, int x1, int y1, int f2, int x2, int y2) {
        // 多层
        if (f1<f2) return isLinked(f2,x2,y2,f1,x1,y1);
        if (f1!=f2)
            return isLinked(f1,x1,y1,f1,stair[f1][2],stair[f1][3])
                && isLinked(f1-1,stair[f1-1][0],stair[f1-1][1],f2,x2,y2);

        if (x1==x2 && y1==y2) return true;
        boolean[][] visited=new boolean[row][col];
        visited[x1][y1]=true;

        Queue<Integer> queue=new LinkedList<>();
        queue.offer(x1); queue.offer(y1);
        while (!queue.isEmpty()) {
            int x=queue.poll(), y=queue.poll();
            for (int[] dir: new int[][] {{-1,0},{0,1},{1,0},{0,-1}}) {
                int nx=x+dir[0], ny=y+dir[1];
                if (nx<0 || nx>=row || ny<0 || ny>=col) continue;
                if (nx==x2 && ny==y2) return true;
                if (visited[nx][ny] || (map[f1][nx][ny]!=ROAD && map[f1][nx][ny]!=UPSTAIR && map[f1][nx][ny]!=DOWNSTAIR)) continue;
                visited[nx][ny]=true;
                queue.offer(nx); queue.offer(ny);
            }
        }

        return false;
    }

    public void run() {
        System.out.println("\n------ Starting BFS ------");
        State state=new State(this, list.get(0));
        State answer=null;

        int index=0;

        HashSet<String> set=new HashSet<>();
        HashMap<String, Integer> map=new HashMap<>();

        // !!! start bfs !!!!!

        long start=System.currentTimeMillis();

        PriorityQueue<State> queue=new PriorityQueue<>((o1,o2)->{
            if (o1.cnt==o2.cnt) return o2.getScore()-o1.getScore();
            return o1.cnt-o2.cnt;
        });
        queue.offer(state);

        while (!queue.isEmpty()) {
            state=queue.poll();

            if (!set.add(state.hashString())) continue;

            if (state.shouldStop()) {
                if (answer==null || answer.getScore()<state.getScore())
                    answer=state;
                continue;
                //break;
            }

            // extend
            for (Node node: state.current.linked) {
                // visited
                if (state.visited[node.id]) continue;

                /*
                // should extend?
                boolean shouldExtend=false;
                for (Node x: node.linked) {
                    if (!x.equals(state.current) && !state.current.linked.contains(x)) {
                        shouldExtend=true;
                        break;
                    }
                }
                if (!shouldExtend) {
                    continue;
                }
                */

                // extend
                State another=new State(state).merge(node);
                if (another==null) continue;
                String hash=another.hashString();
                if (map.getOrDefault(hash, -1)>another.getScore()) continue;
                map.put(hash, another.getScore());
                queue.offer(another);

            }

            index++;
            if (index%1000==0) {
                System.out.println(String.format("Calculating... %d calculated, %d still in queue.", index, queue.size()));
            }

        }
        System.out.println("cnt: "+index);
        System.out.println("------ BFS FINISHED ------");

        if (answer==null) {
            System.out.println("No solution!");
        }
        else {
            for (String string: answer.route) {
                System.out.println(string);
            }
        }

        long end=System.currentTimeMillis();

        System.err.println(String.format("Time used: %.3fs", (end-start)/1000.0));
    }
}

File: AI-cpp/graph.cpp
#include "graph.h"

bool CompareState::operator() (State* s1, State* s2) {
  if (s1->cnt == s2->cnt) return s1->getScore() < s2->getScore();
  return s1->cnt > s2->cnt;
}

Graph::Graph(bool _shouldMerge, bool _shouldEat) {
  scanf("%d%d%d", &floor, &row, &col);
  for (int i=0;i<floor;i++) for (int j=0;j<4;j++) stair[i][j] = -1;
  for (int i=0;i<floor;i++) {
    for (int j=0;j<row;j++) {
      for (int k=0;k<col;k++) {
        scanf("%d", &mp[i][j][k]);
      }
    }
  }
  scanf("%d%d%d%d%d%d%d%d%d", &p_atk, &p_def, &p_mdef,
         &p_red, &p_blue, &p_yellow, &p_green, &p_sword, &p_shield);

  int num;
  scanf("%d", &num);
  for (int i=0;i<num;i++) {
    int id, hp, atk, def, money, special;
    scanf("%d%d%d%d%d%d", &id, &hp, &atk, &def, &money, &special);
    // monsterMap.insert(make_pair(id, new Monster(id, hp, atk, def, money, special)));
    monsterMap[id] = new Monster(id, hp, atk, def, money, special);
  }

  printf("201: %s", monsterMap[201]->toString().c_str());

  int s_start, s_delta, s_hp, s_atk, s_def, s_mdef;
  scanf("%d%d%d%d%d%d", &s_start, &s_delta, &s_hp, &s_atk, &s_def, &s_mdef);
  shop = new Shop(s_start, s_delta, s_hp, s_atk, s_def, s_mdef);

  int hp, atk, def, mdef, money, yellow, blue, red, floor, x, y;
  scanf("%d%d%d%d%d%d%d%d%d%d%d", &hp, &atk, &def, &mdef, &money,
         &yellow, &blue, &red, &floor, &x, &y);

  startNode = (new Node(0, floor, x, y))->setHero(
      new Hero(hp, atk, def, mdef, money, yellow, blue, red, 0));

  nodes.insert(make_pair(startNode->id, startNode));

  this->_shouldMerge = _shouldMerge;
  this->_shouldEat = _shouldEat;

  print_every = 1000;
}

void Graph::initGraph() {
  buildMap();

  if (_shouldMerge)
    mergeNode();

  int index=0;
  for (const auto& nd: nodes) {
    if (nd.second->type == MONSTER_BOSS) bossId = index;
    nodeToIndex.insert(make_pair(nd.first, index++));
  }

}

void Graph::buildMap() {
  for (int i=0;i<floor;i++) {
    for (int j=0;j<row;j++) {
      for (int k=0;k<col;k++) {
        Node* node= nullptr;
        if (mp[i][j][k]==UP_STAIR) {
          stair[i][0]=j; stair[i][1]=k;
        }
        if (mp[i][j][k]==DOWN_STAIR) {
          stair[i][2]=j; stair[i][3]=k;
        }
        if (mp[i][j][k]==YELLOW_KEY)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setYellow(1));
        if (mp[i][j][k]==BLUE_KEY)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setBlue(1));
        if (mp[i][j][k]==RED_KEY)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setRed(1));
        if (mp[i][j][k]==GREEN_KEY)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setGreen(1));
        if (mp[i][j][k]==RED_JEWEL)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setAtk(p_atk));
        if (mp[i][j][k]==BLUE_JEWEL)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setDef(p_def));
        if (mp[i][j][k]==GREEN_JEWEL)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setMDef(p_mdef));
        if (mp[i][j][k]==RED_POTION)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setHp(p_red));
        if (mp[i][j][k]==BLUE_POTION)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setHp(p_blue));
        if (mp[i][j][k]==YELLOW_POTION)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setHp(p_yellow));
        if (mp[i][j][k]==GREEN_POTION)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setHp(p_green));
        if (mp[i][j][k]==SWORD)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setAtk(p_sword));
        if (mp[i][j][k]==SHIELD)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setDef(p_shield));
        if (mp[i][j][k]==SHOP)
          node=(new Node(mp[i][j][k],i,j,k))->setItem((new Item())->setSpecial(1));

        if (mp[i][j][k]==YELLOW_DOOR)
          node=(new Node(mp[i][j][k],i,j,k))->setDoor(1);
        if (mp[i][j][k]==BLUE_DOOR)
          node=(new Node(mp[i][j][k],i,j,k))->setDoor(2);
        if (mp[i][j][k]==RED_DOOR)
          node=(new Node(mp[i][j][k],i,j,k))->setDoor(3);
        if (mp[i][j][k]==GREEN_DOOR)
          node=(new Node(mp[i][j][k],i,j,k))->setDoor(4);
        if (mp[i][j][k]>=MONSTER_BOUND) {
          // Monster* monster=monsterMap.get(map[i][j][k]);
          auto monster = monsterMap[mp[i][j][k]];
          if (monster==nullptr) continue;
          node=(new Node(mp[i][j][k],i,j,k))->setMonster(monster);
        }

        if (node!= nullptr)
          nodes.insert(make_pair(node->id, node));
      }
    }
  }

  // build graph
  for (const auto& i: nodes) {
    for (const auto& j: nodes) {
      if (i.first == j.first) continue;
      Node* n1=i.second, *n2=j.second;
      if (isLinked(n1->f, n1->x, n1->y, n2->f, n2->x, n2->y)) {
        n1->addNode(n2);
        n2->addNode(n1);
      }
    }
  }
}

void Graph::mergeNode() {
  for (const auto& i1: nodes) {
    for (const auto& i2: nodes) {
      if (i1.first==i2.first) continue;
      Node *n1=i1.second, *n2=i2.second;
      if (shouldMerge(n1, n2)) {
        n1->merge(n2);
        delete n2;
        nodes.erase(i2.first);
        mergeNode();
        return;
      }
    }
  }
}

bool Graph::shouldMerge(Node* n1, Node* n2) {
  if (n1->linked.count(n2->id)==0 || n2->linked.count(n1->id)==0) return false;
  if (n1->item!=nullptr && n2->item!=nullptr) return true;
  if (n1->item!=nullptr || n2->item!=nullptr) return false;

  if (n1->type==MONSTER_BOSS || n2->type==MONSTER_BOSS) return false;
  for (const auto& nd: n2->linked) {
    if (n1->linked.count(nd)>0)
      return false;
  }
  for (const auto& x: n1->linked) {
    for (const auto& y: n1->linked) {
      if (x==y || x==n2->id || y==n2->id) continue;
      if (nodes[x]->linked.count(y)==0 || nodes[y]->linked.count(x)==0) return false;
    }
  }
  for (const auto& x: n2->linked) {
    for (const auto& y: n2->linked) {
      if (x==y || x==n1->id || y==n1->id) continue;
      if (nodes[x]->linked.count(y)==0 || nodes[y]->linked.count(x)==0) return false;
    }
  }
  return true;
}

bool Graph::isLinked(int f1, int x1, int y1, int f2, int x2, int y2) {
  if (f1<f2) return isLinked(f2, x2, y2, f1, x1, y1);
  if (f1!=f2)
    return isLinked(f1,x1,y1,f1,stair[f1][2],stair[f1][3])
        && isLinked(f1-1,stair[f1-1][0],stair[f1-1][1],f2,x2,y2);

  if (x1==x2 && y1==y2) return true;

  bool visited[21][21] = {false};
  visited[x1][y1]=true;
  queue<int> q;
  q.push(x1);
  q.push(y1);
  int dir[4][2] = {{-1,0},{0,1},{1,0},{0,-1}};
  while (!q.empty()) {
    int x=q.front(); q.pop();
    int y=q.front(); q.pop();
    for (auto& i : dir) {
      int nx=x+i[0], ny=y+i[1];
      if (nx<0 || nx>=row || ny<0 || ny>=col) continue;
      if (nx==x2 && ny==y2) return true;
      if (visited[nx][ny] || (mp[f1][nx][ny]!=ROAD
          && mp[f1][nx][ny]!=UP_STAIR && mp[f1][nx][ny]!=DOWN_STAIR))
        continue;

      visited[nx][ny]=true;
      q.push(nx); q.push(ny);
    }
  }
  return false;
}

void Graph::print() {
  printf("\n------ Printing Nodes ------\n");
  int valid = 0;
  for (const auto& node: nodes) {
    printf("%s", node.second->toString().c_str());
    if (node.second->doors!=nullptr || node.second->monsters!= nullptr) {
      valid++;
    }
  }
  printf("%d/%lu nodes in total.\n", valid, nodes.size());

}

void Graph::run() {
  printf("\n------ Starting BFS ------\n");

  clock_t begin = clock();

  auto state = new State(this, new Node(startNode));

  int score = 0;
  char answer[10000];

  int index = 0, solutions = 0;

  unordered_set<int64> set;
  unordered_map<int64, int> map;

  priority_queue<State*, vector<State*>, CompareState> pq;

  pq.push(state);

  while (!pq.empty()) {
    state = pq.top();
    pq.pop();

    int64 hash = state->hash();
    if (set.count(hash)>0) {
      delete state;
      continue;
    }

    set.insert(hash);

    if (state->shouldStop()) {
      if (score < state->getScore()) {
        score = state->getScore();
        strcpy(answer, state->route);
      }
      solutions++;
      delete state;
      continue;
    }

    // extend
    for (const auto& node: state->current->linked) {
      if (state->visited[nodeToIndex[node]]) continue;

      /*
      // should extend?
      bool shouldExtend = false;
      for (const auto& x: node->linked) {
        if (x!=state->current && state->current->linked.count(x)==0) {
          shouldExtend = true;
          break;
        }
      }
      if (!shouldExtend)
        continue;
      */

      Node* tmp = state->tryMerge(nodes[node]);
      if (tmp == nullptr) continue;
      State* another = (new State(state, tmp))->merge(nodes[node]);
      // check map
      int& v = map[another->hash()];
      if (v>another->getScore()) {
        delete another;
        continue;
      }
      v = another->getScore();

      pq.push(another);
    }

    delete state;

    index++;
    if (index % print_every == 0) {
      printf("Calculating... %d calculated, %lu still in queue.\n", index, pq.size());
    }
  }
  printf("cnt: %d; solutions: %d\n", index, solutions);
  printf("------ BFS FINISHED ------\n");

  if (strlen(answer)==0) {
    printf("No solution!\n");
  }
  else {
    printf("%s\n", answer);
  }

  clock_t end = clock();
  double seconds = double(end-begin) / CLOCKS_PER_SEC;
  printf("Time used: %.3fs\n", seconds);

}

Graph::~Graph() {
  delete shop;
  for (const auto& nd: nodes) {
    delete nd.second;
  }
  for (const auto& mon: monsterMap) {
    delete mon.second;
  }
}
